CppCoreGuidelines.md

```diff
@@ -7940,7 +7940,37 @@ Casting to a reference expresses that you intend to end up with a valid object,
 
 ##### Example
 
-    ???
+        class Base {
+        public:
+            virtual void f() { }
+        };
+        
+        class Derived : public Base {
+        public:
+            void f() override { }
+        };
+        
+        class NotDerivedFromBase {
+        public:
+            virtual void f() { }
+        };
+        
+        class DerivedFromNotDerivedFromBaseAndFromBase : public NotDerivedFromBase, public Base {
+            void f() override { }
+        };
+        
+        void some_function(NotDerivedFromBase a) {
+          try {
+            Base base = dynamic_cast<Base&>(a);
+          } catch (exception e) {
```

Comment:
catching an exception where it was thrown defeats the point: E.17 and E.18 touch on that a bit, and here it's just always worse than casting to Base* and checking for null.

Reply:
A little off topic, but catching an exception where it is thrown is useful in cases where you don't want to (or can't) leak exceptions (a function is declared `noexcept`) and/or you are fully intending on handling it right there, for example, failed calls to the `std::stoi` family of functions.

On topic-ly, [C.147](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c147-use-dynamic_cast-to-a-reference-type-when-failure-to-find-the-required-class-is-considered-an-error) specifically says to `dynamic_cast<A&>` when you *want* to throw; however, I agree that *in this specific example* catching immediately defeats the purpose of demonstrating the advice.

Reply:
I actually did it by intend. To show what is the difference between the behaviors for C.147 vs C.148. That in one case we would get an exception and in the other we have to check for nullptr. And it's much easier if the examples are similar as much as possible. And the error handling is't the main point here. I can easily have a default behavior that I need to instantiate inside of catch block and assign it to **base** variable if this function is very critical or inside of the transaction that is not allowed to through. Anyway, it wasn't a point, try-catch could be anywhere you want it to be but to emphasize the root cause of the error I prefer to remain my example as is. Otherwise what the purpose of the example if not emphasize the particular problem and make everything else as boilerplate code.

Reply:
I, actually, can put a comment instead of try-catch block if it would be a good compromise of our disagreement =)

Reply:
Replacing the try-catch block with the comment `throws std::bad_cast` is sufficient for me.

Reply:
@cugone, yes it's off-topic, because if this function doesn't want to exit via an exception then it should just use the non-throwing form of `dynamic_cast`, which is exactly what was suggested by the comment you're replying to. In **this** code there is no reason to use the throwing form, it can only be slower.

Reply:
@jwakely The code being commented on is intended to demonstrate the *throwing version*. Why should the non-throwing version be used?

Reply:
Because using the throwing version and catching it makes no sense, and is not a useful example. As already suggested above, simply removing the `try-catch` works fine. Not catching it, or catching and re-throwing, would also be OK. But using the throwing version _and_ catching it right away is a poor example.

You commented to say "catching an exception where it is thrown is useful [in cases that aren't relevant here]" which isn't relevant here. The point is to make the example useful, not discuss the merits of other situations. If you want to discuss E.17 and E.18 please take it to a separate issue, not here.